### DISCUSSION TITLE
Simple way to integrate `mini.completion` and `vim.snippet.expand`?

### DISCUSSION THREAD

**Discussion by Tomcat-42** [nvim-mini/mini.nvim#1414] [Q&A] [answered]
First of all, @echasnovski did an incredible job, and `mini.nvim` is awesome—I’ve fallen in love with it!

I’m using `mini.completion` as a completion engine (I was previously using nvim-cmp), and I’m aware that it doesn’t support snippets, but I’m okay with that (I’m not a heavy snippet user). However, some language servers provide function completions as snippets. For example, `zig-language-server`:

![image](https://github.com/user-attachments/assets/f4b59dff-396e-48de-8fb5-0659dd56d24c)

The question is: How can I (if possible) do a simple expansion of those cases using the `vim.snippet.expand` built-in? My use case is simple—I just want to remove those placeholders. I don’t need fancy placeholder jumping or anything like that.

I'm using this process_items function:

```lua
    require('mini.completion').setup({
      window = {
        info = { border = "single" },
        signature = { border = "single" }
      },
      lsp_completion = {
        process_items = function(items, base)
          local res = vim.tbl_filter(
            function(item)
              return vim.startswith(
                ((item.textEdit or {}).newText or item.insertText or item.label or ''), base)
            end, items)
          table.sort(res, function(a, b) return (a.sortText or a.label) < (b.sortText or b.label) end)
          return res
        end
      }
    })
```

**Comment by echasnovski** [comment, marked as answer]
> First of all, @echasnovski did an incredible job, and `mini.nvim` is awesome—I’ve fallen in love with it!

Thanks for the appreciation! It means a lot!

> I’m using mini.completion as a completion engine (I was previously using nvim-cmp), and I’m aware that it doesn’t support snippets, but I’m okay with that (I’m not a heavy snippet user).

Snippets is a big part of my text editing and I can't recommend enough to invest some time in setting them. That's why I was trying really hard to have 'mini.snippets' that is user-friendly, capable, and robust. Its development is in its final stretch, after which I hope to spend time for 'mini.completion' to *finally* have snippet support.

> The question is: How can I (if possible) do a simple expansion of those cases using the `vim.snippet.expand` built-in? My use case is simple—I just want to remove those placeholders. I don’t need fancy placeholder jumping or anything like that.

Take a look at #1277, in particular [this comment for why this happens](https://github.com/echasnovski/mini.nvim/issues/1277#issuecomment-2410539014)) and [this comment for how to fix it](https://github.com/echasnovski/mini.nvim/issues/1277#issuecomment-2468268529)). For a well behaved server this should do what you ask: not insert snippet for a not "Snippet" completion kind (just text).

Does this solve the issue?

**Reply by Tomcat-42** [reply]
Thanks @echasnovski, it actually solved the issue.

> That's why I was trying really hard to have 'mini.snippets' that is user-friendly, capable, and robust. Its development is in its final stretch, after which I hope to spend time for 'mini.completion' to finally have snippet support.

Awesome, we’ll wait patiently for that next work of art 🙇

**Reply by abeldekat** [reply]
I have been switching back and forth between `nvim-cmp` and `blink.cmp` for completion. 

I would love to use `mini.completion` once it has "lsp snippet expansion" support.

Recently I tested `mini.completion` and was unable to continue due to the "tabstops" in the output, both in `lua` and `rust`.

In order to understand the problem I compiled a list of related [issues](https://github.com/abeldekat/nvim_pde/blob/3436b28317b559e58464486d8d71e77b67c95e1c/lua/ak/config/coding/mini_completion.lua#L6). The solution is to disable "snippets in arguments" at the `lsp` level: In for example `lua_ls`, change `completion.callSnippet` to `Disable` instead of `Replace`. That means I would also need to tweak the rust language server.

Recently I discovered a very handy appliance regarding these `lsp` generated tabstops inside the argument list of function. In `rust`, most statements need to end with a `;`. My approach was to fill in the arguments, `esc` and `A;`. Cumbersome and easy to forget. Solution: After argument list, jump to hidden tabstop `$0` by pressing `ctrl-l` to add the `;`. 

I will also patiently wait for the next work of art!

**Reply by echasnovski** [reply]
> I will also patiently wait for the next work of art!

Well, 'mini.snippets' is out :) The #886 is next after some small backlog cleaning.

**Reply by abeldekat** [reply]
I did notice that...;) Reading the help as we speak!

**Comment by jakubsmolaga** [comment]
Hi @echasnovski, I've been loving `mini` and now that `mini.snippets` is out, I wanted to know if there is any way to integrate it with `mini.completion`. Let's take the following minimal config:
```lua
-- init.lua
require('lspconfig').clangd.setup({})
require('mini.basics').setup()
local gen_loader = require('mini.snippets').gen_loader
require('mini.snippets').setup({ snippets = { gen_loader.from_lang() } })
require('mini.completion').setup()

-- ~/.config/nvim/
-- ├── init.lua
-- └── pack/vendor/start/
--     ├── mini.nvim/
--     │   └── ... // cloned from https://github.com/echasnovski/mini.nvim
--     ├── nvim-lspconfig/
--     │   └── ... // cloned from https://github.com/neovim/nvim-lspconfig
--     └── friendly-snippets/
--         └── ... // cloned from https://github.com/rafamadriz/friendly-snippets
```
With this config I'm getting autocomplete and I'm able to expand snippets with `<C-j>` but I would like to see available snippets in the item list floating window. Is that possible? Any help would be appreciated. Thanks :smile:

**Reply by echasnovski** [reply]
Not yet. The 'mini.completion' part is tracked in #886 and the "see available snippets in completion popup" is planned to be implemented via custom "dummy" in-process LSP server specifically for snippets from 'mini.snippets'.
